core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/sendsemantics/partitioners/ParallelBatchingPartitioner.scala

```diff
@@ -32,13 +73,41 @@ abstract class ParallelBatchingPartitioner(batchSize: Int, receivers: Seq[ActorV
       tuple: ITuple
   ): Option[(ActorVirtualIdentity, DataPayload)] = {
     val index = selectBatchingIndex(tuple)
-    batches(index)(currentSizes(index)) = tuple
-    currentSizes(index) += 1
-    if (currentSizes(index) == batchSize) {
-      currentSizes(index) = 0
-      val retBatch = batches(index)
-      batches(index) = new Array[ITuple](batchSize)
-      return Some((receivers(index), DataFrame(retBatch)))
+    if (Constants.reshapeSkewHandlingEnabled) {
+      // the first receiver in a bucket is the actual receiver for the partition
+      // when there is no mitigation done by Reshape
+      val storedSamples = receiverToWorkloadSamples(bucketsToReceivers(index)(0))
+      storedSamples(currentSampleCollectionIndex) = storedSamples(currentSampleCollectionIndex) + 1
+      tupleIndexForSampling += 1
+      if (tupleIndexForSampling % samplingSize == 0) {
+        if (currentSampleCollectionIndex >= maxSamples - 1) {
+          // Maximum number of samples have been collected.
+          // Write over the older samples
+          currentSampleCollectionIndex = 0
+        } else {
+          if (storedSamples.size < maxSamples) {
+            receiverToWorkloadSamples.keys.foreach(rec => {
+              receiverToWorkloadSamples(rec).append(0)
+            })
+          }
+          currentSampleCollectionIndex += 1
+        }
+        // Set to 0 before starting new sample
+        receiverToWorkloadSamples.keys.foreach(rec =>
+          receiverToWorkloadSamples(rec)(currentSampleCollectionIndex) = 0
+        )
+        tupleIndexForSampling = 0
+      }
+    }
+
+    val receiver: ActorVirtualIdentity = getDefaultReceiverForBucket(index)
+    receiverToBatch(receiver)(receiverToCurrBatchSize(receiver)) = tuple
+    receiverToCurrBatchSize(receiver) += 1
+    if (receiverToCurrBatchSize(receiver) == batchSize) {
+      receiverToCurrBatchSize(receiver) = 0
+      val retBatch = receiverToBatch(receiver)
+      receiverToBatch(receiver) = new Array[ITuple](batchSize)
+      return Some((receiver, DataFrame(retBatch)))
     }
```

Comment:
ok this reshape logic is deeply embedded into `addTupleToBatch` which is a general function. I suggest move this to a separate function to less affect this function.

Reply:
I did realize this and that is why the code in #1414 separates the two,